server/src/main/java/org/elasticsearch/cluster/coordination/JoinHelper.java

```diff
@@ -126,8 +126,16 @@
             StartJoinRequest::new,
             (request, channel, task) -> {
                 final DiscoveryNode destination = request.getSourceNode();
-                sendJoinRequest(destination, Optional.of(joinLeaderInTerm.apply(request)));
-                channel.sendResponse(Empty.INSTANCE);
+                transportService.connectToNode(destination, ActionListener.wrap(ignored -> {
```

Comment:
I've backed this out as I want to address this differently in a follow-up (I want to move this to sendJoinRequest so that we have logging about failed join attempt etc.)